client/tests/signing/basic.rs

```diff
@@ -0,0 +1,349 @@
+        panic!("{conflict:?}, with {tx_payload:#?}");
+    }
+
+    Ok(())
+}
+
+#[tokio::test]
+async fn ed25519_reference_unlocks() -> Result<()> {
+    let secret_manager =
+        SecretManager::Mnemonic(MnemonicSecretManager::try_from_mnemonic(&Client::generate_mnemonic()?)?);
+
+    let bech32_address_0 = &secret_manager
+        .generate_addresses(SHIMMER_COIN_TYPE, 0, 0..1, false, None)
+        .await?[0]
+        .to_bech32(SHIMMER_TESTNET_BECH32_HRP);
+
+    let protocol_parameters = protocol_parameters();
+
+    let inputs = build_inputs(vec![
+        Basic(
+            1_000_000,
+            bech32_address_0,
+            None,
+            None,
+            None,
+            None,
+            None,
+            Some(Chain::from_u32_hardened(vec![
+                HD_WALLET_TYPE,
+                SHIMMER_COIN_TYPE,
+                0,
+                0,
+                0,
+            ])),
+        ),
+        Basic(
+            1_000_000,
+            bech32_address_0,
+            None,
+            None,
+            None,
+            None,
+            None,
+            Some(Chain::from_u32_hardened(vec![
+                HD_WALLET_TYPE,
+                SHIMMER_COIN_TYPE,
+                0,
+                0,
+                0,
+            ])),
+        ),
+        Basic(
+            1_000_000,
+            bech32_address_0,
+            None,
+            None,
+            None,
+            None,
+            None,
+            Some(Chain::from_u32_hardened(vec![
+                HD_WALLET_TYPE,
+                SHIMMER_COIN_TYPE,
+                0,
+                0,
+                0,
+            ])),
+        ),
+    ]);
+
+    let outputs = build_outputs(vec![Basic(
+        3_000_000,
+        bech32_address_0,
+        None,
+        None,
+        None,
+        None,
+        None,
+        Some(Chain::from_u32_hardened(vec![
+            HD_WALLET_TYPE,
+            SHIMMER_COIN_TYPE,
+            0,
+            0,
+            0,
+        ])),
+    )]);
+
+    let essence = TransactionEssence::Regular(
+        RegularTransactionEssence::builder(
+            protocol_parameters.network_id(),
+            InputsCommitment::new(inputs.iter().map(|i| &i.output)),
+        )
+        .with_inputs(
+            inputs
+                .iter()
+                .map(|i| Input::Utxo(UtxoInput::from(*i.output_metadata.output_id())))
+                .collect(),
+        )
+        .with_outputs(outputs)
+        .finish(&protocol_parameters)?,
+    );
+
+    let prepared_transaction_data = PreparedTransactionData {
+        essence,
+        inputs_data: inputs,
+        remainder: None,
+    };
+
+    let unlocks = secret_manager
+        .sign_transaction_essence(&prepared_transaction_data)
+        .await?;
+
+    assert_eq!(unlocks.len(), 3);
+    assert_eq!((*unlocks).get(0).unwrap().kind(), SignatureUnlock::KIND);
+    assert_eq!((*unlocks).get(1).unwrap().kind(), ReferenceUnlock::KIND);
```

Comment:
Maybe do `matches!` on the unlock to also assert the index ?

Reply:
Can you show an example? Not sure what you mean, isn't `.get(1).unwrap()` basically doing that?

Reply:
I meant something like `assert!(matches!(*unlocks.get(1).unwrap(), ReferenceUnlock(0)));`. But that's actually not possible because the field is not pub. But atm you are not checking the inner index. Maybe add another assert for it ?